hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/avro/HoodieAvroParquetReader.java

```diff
@@ -38,24 +42,61 @@
+import static org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils.constructHiveOrderedSchema;
+import static org.apache.hudi.hadoop.utils.HoodieRealtimeRecordReaderUtils.getNameToFieldMap;
 import static org.apache.parquet.hadoop.ParquetInputFormat.getFilter;
 
 public class HoodieAvroParquetReader extends RecordReader<Void, ArrayWritable> {
 
+  private static final Logger LOG = LoggerFactory.getLogger(HoodieAvroParquetReader.class);
   private final ParquetRecordReader<GenericData.Record> parquetRecordReader;
   private Schema baseSchema;
 
   public HoodieAvroParquetReader(InputSplit inputSplit, Configuration conf) throws IOException {
-    // get base schema
-    ParquetMetadata fileFooter =
-        ParquetFileReader.readFooter(conf, ((ParquetInputSplit) inputSplit).getPath(), ParquetMetadataConverter.NO_FILTER);
-    MessageType messageType = fileFooter.getFileMetaData().getSchema();
-    baseSchema = new AvroSchemaConverter(conf).convert(messageType);
+    Path filePath = ((ParquetInputSplit) inputSplit).getPath();
+    Schema writerSchema;
+    try {
+      HoodieTableMetaClient metaClient = getTableMetaClientForBasePathUnchecked(conf, filePath);
+      writerSchema = new TableSchemaResolver(metaClient).getTableAvroSchema();
+      LOG.warn("Got writer schema from table schema: {}", writerSchema);
+    } catch (Exception e) {
+      LOG.error("Failed to get writer schema from table schema", e);
+      LOG.warn("Falling back to reading writer schema from parquet file");
+      ParquetMetadata fileFooter = ParquetFileReader.readFooter(conf, filePath, ParquetMetadataConverter.NO_FILTER);
+      MessageType messageType = fileFooter.getFileMetaData().getSchema();
+      writerSchema = new AvroSchemaConverter(conf).convert(messageType);
+    }
+    JobConf jobConf = new JobConf(conf);
+    try {
+      // Add partitioning fields to writer schema for resulting row to contain null values for these fields
+      String partitionFields = jobConf.get(hive_metastoreConstants.META_TABLE_PARTITION_COLUMNS, "");
+      List<String> partitioningFields =
+          partitionFields.length() > 0 ? Arrays.stream(partitionFields.split("/")).collect(Collectors.toList())
+              : new ArrayList<>();
+      LOG.warn("Got partitioning fields: {}", partitioningFields);
+      writerSchema = addPartitionFields(writerSchema, partitioningFields);
+      LOG.warn("Added partition fields to writer schema: {}", writerSchema);
+    } catch (Exception e) {
+      LOG.error("Failed to add partition fields to writer schema", e);
+    }
+    Map<String, Schema.Field> schemaFieldsMap = getNameToFieldMap(writerSchema);
+    LOG.warn("Got schema fields map: {}", schemaFieldsMap);
+    baseSchema = constructHiveOrderedSchema(writerSchema, schemaFieldsMap, jobConf.get(hive_metastoreConstants.META_TABLE_COLUMNS, EMPTY_STRING));
+    LOG.warn("Got hive ordered schema: {}", baseSchema);
```

Comment:
Does we have some specific test cases?

Reply:
This is a tentative fix. Need to add tests. I wanted to verify something for Athena.